File: backend/src/graduate_entrance/models/problems.py
from __future__ import annotations

from datetime import UTC, date, datetime
from uuid import UUID, uuid4

from sqlalchemy import (
    JSON,
    Boolean,
    CheckConstraint,
    Date,
    DateTime,
    Float,
    ForeignKey,
    String,
    Text,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from graduate_entrance.db.base import Base


def utc_now() -> datetime:
    return datetime.now(UTC)


class Problem(Base):
    __tablename__ = "problems"
    __table_args__ = (
        CheckConstraint("kind IN ('wrong', 'hard', 'good')", name="ck_problems_kind"),
        CheckConstraint(
            "cause IN ('', 'concept', 'calculation', 'method', 'memory', 'misread', 'other')",
            name="ck_problems_cause",
        ),
        CheckConstraint("status IN ('draft', 'confirmed')", name="ck_problems_status"),
        CheckConstraint("ef >= 1.3", name="ck_problems_ef"),
        CheckConstraint("interval_days >= 0", name="ck_problems_interval_days"),
        CheckConstraint("reps >= 0", name="ck_problems_reps"),
    )

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    subject_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("subjects.id", ondelete="SET NULL"),
        index=True,
    )
    content_md: Mapped[str] = mapped_column(Text, default="")
    images: Mapped[list[str]] = mapped_column(JSON, default=list)
    source_ref: Mapped[str] = mapped_column(String(240), default="")
    kind: Mapped[str] = mapped_column(String(16), default="wrong")
    my_answer_md: Mapped[str] = mapped_column(Text, default="")
    cause: Mapped[str] = mapped_column(String(32), default="")
    note: Mapped[str] = mapped_column(Text, default="")
    status: Mapped[str] = mapped_column(String(16), default="draft", index=True)
    ef: Mapped[float] = mapped_column(Float, default=2.5)
    interval_days: Mapped[int] = mapped_column(default=0)
    due_date: Mapped[date | None] = mapped_column(Date, index=True)
    reps: Mapped[int] = mapped_column(default=0)
    confirmed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
    )

    kp_links: Mapped[list[ProblemKnowledgePoint]] = relationship(
        back_populates="problem",
        cascade="all, delete-orphan",
        order_by="ProblemKnowledgePoint.knowledge_point_id",
    )
    solutions: Mapped[list[Solution]] = relationship(
        back_populates="problem",
        cascade="all, delete-orphan",
        order_by="Solution.created_at",
    )


class ReviewLog(Base):
    __tablename__ = "review_logs"
    __table_args__ = (
        CheckConstraint(
            "grade IN ('forgot', 'vague', 'mastered')",
            name="ck_review_logs_grade",
        ),
    )

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    problem_id: Mapped[UUID] = mapped_column(
        ForeignKey("problems.id", ondelete="CASCADE"),
        index=True,
    )
    grade: Mapped[str] = mapped_column(String(16))
    reviewed_on: Mapped[date] = mapped_column(Date, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now)


class Solution(Base):
    __tablename__ = "solutions"
    __table_args__ = (
        CheckConstraint("source IN ('self', 'answer', 'gpt')", name="ck_solutions_source"),
    )

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    problem_id: Mapped[UUID] = mapped_column(
        ForeignKey("problems.id", ondelete="CASCADE"),
        index=True,
    )
    content_md: Mapped[str] = mapped_column(Text)
    method_tag: Mapped[str] = mapped_column(String(80), default="")
    source: Mapped[str] = mapped_column(String(16), default="self")
    verified: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
    )

    problem: Mapped[Problem] = relationship(back_populates="solutions")


class ProblemKnowledgePoint(Base):
    __tablename__ = "problem_knowledge_points"
    __table_args__ = (
        CheckConstraint(
            "role IN ('primary', 'secondary')",
            name="ck_problem_knowledge_points_role",
        ),
        CheckConstraint(
            "weight > 0 AND weight <= 1",
            name="ck_problem_knowledge_points_weight",
        ),
    )

    problem_id: Mapped[UUID] = mapped_column(
        ForeignKey("problems.id", ondelete="CASCADE"),
        primary_key=True,
    )
    knowledge_point_id: Mapped[UUID] = mapped_column(
        ForeignKey("knowledge_points.id", ondelete="CASCADE"),
        primary_key=True,
        index=True,
    )
    role: Mapped[str] = mapped_column(String(16))
    weight: Mapped[float] = mapped_column(Float)

    problem: Mapped[Problem] = relationship(back_populates="kp_links")

File: web/src/services/problems.ts
import { apiFetch } from './api'

export type ProblemKind = 'wrong' | 'hard' | 'good'
export type ProblemCause =
  | ''
  | 'concept'
  | 'calculation'
  | 'method'
  | 'memory'
  | 'misread'
  | 'other'
export type ProblemStatus = 'draft' | 'confirmed'
export type KnowledgePointRole = 'primary' | 'secondary'

export interface ProblemKnowledgePointInput {
  knowledge_point_id: string
  role: KnowledgePointRole
  weight: number
}

export interface ProblemKnowledgePointRead {
  knowledge_point_id: string
  knowledge_point_name: string
  role: KnowledgePointRole
  weight: number
}

export interface SolutionRead {
  id: string
  content_md: string
  method_tag: string
  source: 'self' | 'answer' | 'gpt'
  verified: boolean
  created_at: string
}

export interface Problem {
  id: string
  subject_id: string | null
  subject_name: string | null
  content_md: string
  images: string[]
  source_ref: string
  kind: ProblemKind
  my_answer_md: string
  cause: ProblemCause
  note: string
  status: ProblemStatus
  due_date: string | null
  reps: number
  confirmed_at: string | null
  created_at: string
  knowledge_points: ProblemKnowledgePointRead[]
  solutions: SolutionRead[]
}

export interface PendingProblems {
  total: number
  problems: Problem[]
}

export type ReviewGrade = 'forgot' | 'vague' | 'mastered'

export interface DueReviews {
  total: number
  as_of: string
  problems: Problem[]
}

export interface ReviewResult {
  problem: Problem
  grade: ReviewGrade
  ef: number
  interval_days: number
  reps: number
  due_date: string
}

export interface ProblemConfirmPayload {
  content_md: string
  kind: ProblemKind
  cause: ProblemCause
  my_answer_md: string
  note: string
  source_ref: string
  knowledge_points: ProblemKnowledgePointInput[]
}

async function requestJson<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await apiFetch(`/api${path}`, {
    ...init,
    headers: {
      'Content-Type': 'application/json',
      ...init?.headers,
    },
  })
  if (!response.ok) {
    throw new Error(`Problems request failed with status ${response.status}`)
  }
  return (await response.json()) as T
}

export interface ProblemSubmitInput {
  subjectId?: string
  kind: ProblemKind
  contentMd: string
  sourceRef: string
  myAnswerMd: string
  note: string
  images: File[]
}

export async function submitProblem(input: ProblemSubmitInput): Promise<Problem> {
  const form = new FormData()
  if (input.subjectId) {
    form.set('subject_id', input.subjectId)
  }
  form.set('kind', input.kind)
  form.set('content_md', input.contentMd)
  form.set('source_ref', input.sourceRef)
  form.set('my_answer_md', input.myAnswerMd)
  form.set('note', input.note)
  for (const image of input.images) {
    form.append('images', image)
  }
  const response = await apiFetch('/api/problems', {
    method: 'POST',
    body: form,
  })
  if (!response.ok) {
    throw new Error(`Problems request failed with status ${response.status}`)
  }
  return (await response.json()) as Problem
}

export function fetchPendingProblems(): Promise<PendingProblems> {
  return requestJson<PendingProblems>('/problems/pending')
}

export function confirmProblem(id: string, payload: ProblemConfirmPayload): Promise<Problem> {
  return requestJson<Problem>(`/problems/${id}/confirm`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export interface ExtractedKnowledgePoint {
  knowledge_point_id: string
  knowledge_point_name: string
  role: KnowledgePointRole
  weight: number
}

export interface ExtractedSolution {
  content_md: string
  method_tag: string
}

export interface ProblemExtractionResult {
  problem_id: string
  model: string
  content_md: string
  knowledge_points: ExtractedKnowledgePoint[]
  solution: ExtractedSolution | null
}

export interface SolutionCreatePayload {
  content_md: string
  method_tag: string
  source: 'self' | 'answer' | 'gpt'
}

export function addSolution(id: string, payload: SolutionCreatePayload): Promise<Problem> {
  return requestJson<Problem>(`/problems/${id}/solutions`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function extractProblem(id: string): Promise<ProblemExtractionResult> {
  return requestJson<ProblemExtractionResult>(`/problems/${id}/extract`, {
    method: 'POST',
  })
}

export function fetchDueReviews(includeDrafts = true): Promise<DueReviews> {
  const query = includeDrafts ? '' : '?include_drafts=false'
  return requestJson<DueReviews>(`/problems/reviews/due${query}`)
}

export function reviewProblem(id: string, grade: ReviewGrade): Promise<ReviewResult> {
  return requestJson<ReviewResult>(`/problems/${id}/review`, {
    method: 'POST',
    body: JSON.stringify({ grade }),
  })
}

export async function fetchProblemImage(name: string): Promise<string> {
  const response = await apiFetch(`/api/problems/images/${encodeURIComponent(name)}`)
  if (!response.ok) {
    throw new Error(`Problems request failed with status ${response.status}`)
  }
  return URL.createObjectURL(await response.blob())
}

export interface KnowledgePointInsight {
  knowledge_point_id: string
  knowledge_point_name: string
  problem_count: number
  weighted_errors: number
  forgot_reviews: number
  total_reviews: number
  weakness_score: number
}

export interface CauseInsight {
  cause: ProblemCause
  count: number
}

export interface SubjectInsight {
  subject_id: string | null
  subject_name: string
  problem_count: number
  wrong_count: number
}

export interface WeeklyTrendPoint {
  week_start: string
  new_problems: number
  reviews: number
  forgot: number
  vague: number
  mastered: number
}

export interface ProblemInsights {
  as_of: string
  total_problems: number
  confirmed_problems: number
  knowledge_points: KnowledgePointInsight[]
  causes: CauseInsight[]
  subjects: SubjectInsight[]
  weekly_trend: WeeklyTrendPoint[]
}

export function fetchProblemInsights(): Promise<ProblemInsights> {
  return requestJson<ProblemInsights>('/stats/insights')
}
